server/src/main/java/org/apache/iotdb/db/engine/storagegroup/DataRegion.java

```diff
@@ -2069,6 +2249,235 @@ private List<TsFileResource> getFileResourceListForQuery(
+          "Alter failed. "
+              + "alter.log detected, other alter operations may be running, please do it later.");
+    }
+    // rewrite target tsfiles
+    boolean done = false;
+    try (AlteringLogger alteringLogger = new AlteringLogger(logFile)) {
+      Set<Long> timePartitions = tsFileManager.getTimePartitions();
+      // Record the ALTER process for server restart recovery
+      alteringLogger.logHeader(fullPath, curEncoding, curCompressionType, timePartitions);
+      for (Long timePartition : timePartitions) {
+        logger.info("[alter timeseries] {} alterDataInTsFiles seq({})", logKey, timePartition);
+        try {
+          rewriteDataInTsFiles(
+              tsFileManager.getSequenceListByTimePartition(timePartition),
+              fullPath,
+              curEncoding,
+              curCompressionType,
+              timePartition,
+              true,
+              alteringLogger,
+              null,
+              logKey);
+          logger.info("[alter timeseries] {} alterDataInTsFiles unseq({})", logKey, timePartition);
+          rewriteDataInTsFiles(
+              tsFileManager.getUnsequenceListByTimePartition(timePartition),
+              fullPath,
+              curEncoding,
+              curCompressionType,
+              timePartition,
+              false,
+              alteringLogger,
+              null,
+              logKey);
+        } catch (IOException e) {
+          /**
+           * TODO If an exception occurs in the operation of a single tsfile, you need to terminate
+           * the operation and return 1. The modified schema and rewritten tsfile need to provide
+           * rollback commands 2. The tsfile that has not been rewritten needs to provide a recovery
+           * command
+           */
+          logger.error(
+              "[alter timeseries] " + logKey + " timePartition " + timePartition + " error", e);
+          throw e;
+        }
+      }
+      done = true;
+    } catch (Exception e) {
+      logger.error("[alter timeseries] " + logKey + " error", e);
+      throw e;
+    } finally {
+      logger.info("[alter timeseries] {} rewriteUnlock", logKey);
+      tsFileManager.rewriteUnlock();
+      // The process is complete and the logFile is deleted
+      if (done && logFile.exists()) {
+        FileUtils.delete(logFile);
+      }
+    }
+  }
+
+  private void rewriteDataInTsFiles(
+      List<TsFileResource> tsFileList,
+      PartialPath fullPath,
+      TSEncoding curEncoding,
+      CompressionType curCompressionType,
+      long timePartition,
+      boolean sequence,
+      AlteringLogger alteringLogger,
+      Set<TsFileIdentifier> undoneFiles,
+      String logKey)
+      throws IOException {
+
+    if (tsFileList == null || tsFileList.isEmpty()) {
+      return;
+    }
+    // log timePartition start
+    alteringLogger.startTimePartition(tsFileList, timePartition, sequence);
+    for (TsFileResource tsFileResource : tsFileList) {
```

Comment:
It seems that all TsFiles will be rewritten no matter it has corresponding device or not. Add a judgment to filter out tsfiles that do not contain the corresponding device.